parsing/pprintast.ml

```diff
@@ -397,22 +397,34 @@ and core_type1 ctxt f x =
 (********************pattern********************)
 (* be cautious when use [pattern], [pattern1] is preferred *)
 and pattern ctxt f x =
-  let rec list_of_pattern acc = function (* only consider ((A|B)|C)*)
-    | {ppat_desc= Ppat_or (p1,p2); ppat_attributes = []} ->
-        list_of_pattern  (p2::acc) p1
-    | x -> x::acc
-  in
   if x.ppat_attributes <> [] then begin
     pp f "((%a)%a)" (pattern ctxt) {x with ppat_attributes=[]}
       (attributes ctxt) x.ppat_attributes
   end
   else match x.ppat_desc with
     | Ppat_alias (p, s) ->
-        pp f "@[<2>%a@;as@;%a@]" (pattern ctxt) p protect_ident s.txt (* RA*)
-    | Ppat_or _ -> (* *)
-        pp f "@[<hov0>%a@]" (list ~sep:"@,|" (pattern ctxt))
-          (list_of_pattern [] x)
-    | _ -> pattern1 ctxt f x
+        pp f "@[<2>%a@;as@;%a@]"
+          (* we do not use pattern_or here: it would be correct
```

Comment:
Using `pattern` like before would be correct too, since `_ as x as y` is valid.

Reply:
Yes, `pattern` was incorrect before due to the pretense that `as` and `|` had the same priority. Now that the two levels are split it would be correct again (this was the main point of the change). I am not completely sure which is better. `pattern` has a more standard structure, but I find the printing for `A | B as x` very confusing.

Note: if we wanted to avoid the parentheses in `(A as x) | B`, we would need a different, wicked change, namely handling another layer of Ppat_alias inside the leftmost or-pattern (this would not be `pattern`, but a rule that allows `alias` and is below `pattern_or`.)

Reply:
On the second thought, I may go with `pattern` again. I like that it lets me remove a comment, and also `A | B as x` has the upper-side that (although it is confusing) the "natural" interpretation is ill-scoped, so people have to understand its meaning to be able to interpret it at all.